metrics/sinks/gcm/gcm.go

```diff
@@ -321,21 +324,39 @@ func CreateGCMSink(uri *url.URL) (core.DataSink, error) {
 		return nil, fmt.Errorf("invalid metrics parameter: %s", metrics)
 	}
 
-	if err := gce_util.EnsureOnGCE(); err != nil {
-		return nil, err
-	}
-
-	// Detect project ID
-	projectId, err := gce.ProjectID()
-	if err != nil {
-		return nil, err
-	}
+	var projectId string
+	var gcmService *gcm.Service
 
-	// Create Google Cloud Monitoring service.
-	client := oauth2.NewClient(oauth2.NoContext, google.ComputeTokenSource(""))
-	gcmService, err := gcm.New(client)
-	if err != nil {
-		return nil, err
+	if os.Getenv(gcpCredentialEnv) != "" {
+		// GCP credentials supplied, so assume we're not in GCE
+		if os.Getenv(gcpProjectEnv) == "" {
+			return nil, fmt.Errorf("no GCP project ID supplied")
```

Comment:
There is another option -- the credentials file actually has the project id embedded in it. If the user didn't supply one, why not try the one from the credentials file, and then the one from the metadata server, in that order?